DM-26629: switch to calibration collections and remove calibration_label

TallJimbo opened the pull request.

TallJimbo committed:
Add placeholder queryDatasetAssociations.

This will ultimately need more flexibility in how to accept dataset
constraints, rather than just pulling down everything in a dataset
type + collection combination, but that is blocked by DM-26692.

TallJimbo committed:
Support collections in import/export.

TallJimbo committed:
Only query RUN collections for datasets when asked to query all.

As all datasets are in a RUN collection, this both avoids duplicates
and prevents triggering the limitation on CALIBRATION collections in
queryDatasets in this common case.

TallJimbo committed:
Re-enable mypy on transfers code and fix breakage.

This was accidentally disabled by moving the code out of core.

TallJimbo committed:
Remove calibration_label dimension.

TallJimbo committed:
Remove do-nothing --output_run option from 'butler import'.

The Butler.import_ method never uses the run passed to the Butler at
construction, and hence it makes no sense as a command-line argument,
let alone a required one.  Imported datasets always land in the
collections specified in the export file.

TallJimbo committed:
Skip CALIBRATION collections when querying non-calibration datasets.

TallJimbo committed:
Add direct variant of Butler.getDeferred.

TallJimbo committed:
Document transfer='auto' behavior for PosixDatastore.

timj commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -664,12 +664,44 @@ def getDirect(self, ref: DatasetRef, *, parameters: Optional[Dict[str, Any]] = N
+        obj : `DeferredDatasetHandle`
+            A handle which can be used to retrieve a dataset at a later time.
+
+        Raises
```
I don't think these docstrings are relevant.

timj commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -664,12 +664,44 @@ def getDirect(self, ref: DatasetRef, *, parameters: Optional[Dict[str, Any]] = N
+        TypeError
+            Raised if no collections were provided.
+        """
+        return DeferredDatasetHandle(butler=self, ref=ref, parameters=parameters)
```
In the interests of failing early, can we check that the supplied ref is actually resolved?

timj commented on python/lsst/daf/butler/configs/datastores/webdavDatastore.yaml:
```diff
@@ -8,6 +8,6 @@ datastore:
     # valid_first and valid_last here are YYYYMMDD; we assume we'll switch to
     # MJD (DM-15890) before we need more than day resolution, since that's all
     # Gen2 has.
-    default: "{run:/}/{datasetType}.{component:?}/{tract:?}/{patch:?}/{label:?}/{abstract_filter:?}/{subfilter:?}/{physical_filter:?}/{visit:?}/{datasetType}_{component:?}_{tract:?}_{patch:?}_{label:?}_{abstract_filter:?}_{physical_filter:?}_{calibration_label:?}_{visit:?}_{exposure:?}_{detector:?}_{instrument:?}_{skymap:?}_{skypix:?}_{run}"
```
Yes, I should have a shared templates.yaml for these three datastores.

timj commented on python/lsst/daf/butler/core/datasets/association.py:
```diff
@@ -0,0 +1,49 @@
+    __slots__ = ("ref", "collection", "timespan")
+
+    ref: DatasetRef
+    """Reference to a dataset.
```
This can be resolved or unresolved?

TallJimbo replied:
Always resolved.  Will document.

timj commented on python/lsst/daf/butler/registry/_registry.py:
```diff
@@ -1497,6 +1500,72 @@ def queryDimensionRecords(self, element: Union[DimensionElement, str], *,
+        removed in the future, and should be avoided in production code
+        whenever possible.
+
+        Parameters
```
Missing some parameters here.

TallJimbo replied:
I think I'd intended to delete both of those from the signature as well, but actual usage reveals that I do need `flattenChains` (but `includeChains` definitely doesn't make sense for this method).  Will fix.

timj commented on python/lsst/daf/butler/registry/_registry.py:
```diff
@@ -1497,6 +1500,72 @@ def queryDimensionRecords(self, element: Union[DimensionElement, str], *,
+                if collectionRecord.type is CollectionType.CALIBRATION:
+                    timespan = tsRepr.extract(row)
+                else:
+                    timespan = None
```
I really thought that mypy would get cranky about timespan here being a different type to the one two lines before. I always had to rewrite those as:

```python
timespan = tsRepr.extract(row) if collectionRecord.type is CollectionType.CALIBRATION else None
```
for mypy to work out that timespan can be `Optional[Timespan]`. Maybe I'm misremembering.

TallJimbo replied:
Yup, I have the same recollection, and I'm not sure if mypy has gotten smarter or if this case is easier in a non-obvious way.  I do expect mypy will eventually get smart enough to infer the optional in essentially all cases like this, so I'm not inclined to fight it when it seems to.

timj commented on python/lsst/daf/butler/registry/queries/_builder.py:
```diff
@@ -143,6 +143,13 @@ def joinDataset(self, datasetType: DatasetType, collections: Any, *,
         else:
             collections = CollectionQuery.fromExpression(collections)
+        # If we are searching all collections with no constraints, loop over
+        # RUN collections only, because that will include all datasets.
```
Very good. Obvious in retrospect but I'm sure it all helps.

timj commented on python/lsst/daf/butler/registry/wildcards.py:
```diff
@@ -839,6 +839,12 @@ def iter(
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, CollectionQuery):
+            return self._search == other._search and self._patterns == other._patterns
+        else:
```
The else doesn't really do much so could be dropped and just have the return as the last line in the method.

TallJimbo replied:
I've seen style guides (in other languages, I think?) that recommend this form to "preserve the symmetry of the branching" or some such, and I have never had a good sense for whether one form other the other was idiomatic in Python.  But if this was strange enough to you to merit a comment, it probably merits changing.

timj commented on python/lsst/daf/butler/transfers/_context.py:
```diff
@@ -79,8 +81,35 @@ def __init__(self, registry: Registry, datastore: Datastore, backend: RepoExport
+
+        Exporting a `~CollectionType.TAGGED` or `~CollectionType.CALIBRATION`
+        collection will cause its associations with exported datasets to also
+        be exported, but it does not export those datasets automatically.
```
so you can end up with an empty calibration collection after import? Or a broken one because the references exist but the datasets don't?

TallJimbo replied:
Empty is indeed possible (I think that's unavoidable if a partial export of a collection is allowed).  Broken should not be possible.

timj commented on python/lsst/daf/butler/transfers/_context.py:
```diff
@@ -207,13 +238,106 @@ def _finish(self) -> None:
+        while chains:
+            unblocked = {
+                parent for parent, children in chains.items()
+                if not any(child in chains.keys() for child in children)
```
is the keys() needed here? Normally `__contains__` would work on the dict directly.

TallJimbo replied:
Not needed, I just thought it was easier to read.  I know `.keys` is basically always visual noise for some people, but when it's not extremely obvious that the target of `__contains__` or `__iter__` is a mapping I find it very helpful to have it called out.

timj replied:
Ok, it looks odd to me because I'm then wondering what's special about the `keys` that means that `in` doesn't work (and surely it makes python work harder since it then has to generate keys and do the `__contains__` on that?)

timj reviewed (approved):
Looks good. A few minor comments and questions.

TallJimbo committed:
Include DatasetType.isCalibration() in __eq__, __repr__.

This helps us spot and report inconsistent definitions in
PipelineTasks.

TallJimbo committed:
Support calibration indirection directly in butler.get.

I'm not certain this is the right place for this logic, but the
huge simplification it provides for ci_cpp_gen3 tests (which will carry
over directly to interactive CPP validation work) suggests that it
belongs somewhere in daf_butler.
